### DISCUSSION TITLE
Edge Side Rendering (ESR) with Cloudflare Workers

### DISCUSSION THREAD

**Discussion by iamacarpet** [inertiajs/inertia#1899] [Ideas]
As an idea, we are floating around the option of introducing Edge Side Rendering (ESR) on Cloudflare Workers.

What if we want the benefits of SSR, without needing the NodeJS runtime & the associated CPU usage server side?

Cloudflare Workers provides a JavaScript runtime that proxies requests to an origin, while allowing requests and HTML DOM to be modified before being returned to the client - this means it should be possible to:

- Fetch a response from the origin that is configured for CSR
- Parse & inspect the returned HTML DOM to pull out the `<div id="app" data-page="` data.
- Run `ReactDOMServer.renderToString`, ideally using a ESR specific JS bundle containing the relevant `jsx` data.
- Re-write the results into the HTML & return the response to the client, caching as necessary.
- Client would get HTML to match a page generated by SSR, and can hydrate to become interactive.

The main challenge I foresee at the moment is the need for an ESR specific bundle to be loaded by the worker, so it can proxy any application running InertiaJS and doesn't need to be specifically deployed with the application data - it'll request the JS bundle with the `jsx` data, as a client would to perform CSR.

I'm not asking for this to be implemented for us - if we go ahead with the idea, I'll be happy to take it towards a proof-of-concept - however, what is the community appetite like for something like this?

Is it something anyone else would find useful? And/or, would you consider accepting it upstream?

As an example architecture:

![laravel-inertiajs-esr](https://github.com/inertiajs/inertia-laravel/assets/1333857/66df4cd9-9f5d-4305-ac55-f56a3e93d65e)

As opposed to either CSR:

![laravel-inertiajs-csr](https://github.com/inertiajs/inertia-laravel/assets/1333857/eb5b00a7-5c60-419c-8552-45d8c9a133a5)

Or traditional SSR:

![laravel-inertiajs-ssr](https://github.com/inertiajs/inertia-laravel/assets/1333857/5c8f4961-1c5f-4bac-8dc4-45bca36e5373)

**Comment by reinink** [comment]
Hey! Thanks for sharing this idea! Given that it's an idea and not an issue I've converted this to a discussion.

Using Cloudflare workers to do SSR instead of needing to run Node on your own server is definitely appealing. I don't understand all the nuances of doing this though, so really don't have a great gut sense of what kind of changes would be required in Inertia to support this.

If there's a couple simple changes that could be made to Inertia to support this, that's cool, but I'm not sure we want to require a bunch of Cloudflare specific implementation code in Inertia itself. To me that would be better handled via some type of Cloudflare + Inertia plugin.

Also, just being honest, I don't personally have a lot of time to invest in this particular problem right now. We have a number of more pressing improvements we want to make to Inertia core. But if you wanted to put together a PR outlining how to do this (ideally with the minimum number of changes possible), I'd be happy to have a peek at it. Just fair warning: there's a decent chance it doesn't get merged in if it feels at all too complicated, as it would then become code that we're responsible for maintaining.

Hope that makes sense! 🤙

**Comment by tobimori** [comment]
Consider me interested

**Comment by sevfurneaux** [comment]
Were there any further developments on this, @iamacarpet? This sounded like an interesting technique.

**Comment by JohnDotOwl** [comment]
Very interesting, i think edge cache / rendering is the missing gap for laravel given how the default starter kits are mostly react/vue based.
